torch/_refs/linalg/__init__.py

```diff
@@ -0,0 +1,129 @@
+            lambda: "{fn_name}: dtype should be {d} for {d} inputs. Got {dtype}".format(
+                fn_name=fn_name,
+                d="complex" if is_complex_dtype(x_dtype) else "real",
+                dtype=dtype,
+            ),
+        )
+        check(
+            get_higher_dtype(dtype, x_dtype) == dtype,
+            lambda: f"{fn_name}: the dtype of the input ({x_dtype}) should be convertible "
+            "without narrowing to the specified dtype ({dtype})",
+        )
+
+
+@out_wrapper(exact_dtype=True)
+def vector_norm(
+    x: TensorLikeType,
+    ord: float = 2.0,
+    dim: Optional[DimsType] = None,
+    keepdim: bool = False,
+    *,
+    dtype: Optional[torch.dtype] = None,
+) -> Tensor:
+    # Checks
+    linalg.utils.check_fp_or_complex(
+        x.dtype, "linalg.vector_norm", allow_low_precision_dtypes=True
+    )
+
+    if isinstance(dim, int):
+        dim = [dim]  # type: ignore[assignment]
+    elif not isinstance(dim, List) and dim is not None:
+        # refs.sum just accepts List rather than DimType
+        dim = list(dim)  # type: ignore[assignment]
+
+    if x.numel() == 0 and (ord < 0.0 or ord == float("inf")):
+        check(
+            dim is not None and len(dim) != 0,
+            lambda: f"linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
+            "because the operation does not have an identity",
+        )
+        shape = x.shape
+        assert dim is not None  # mypy does not seem to be able to see through check?
+        for d in dim:
+            check(
+                shape[d] != 0,
+                lambda: f"linalg.vector_norm cannot compute the {ord} norm on the "
+                f"dimension {d} because this dimension is empty and the "
+                "operation does not have an identity",
+            )
+    check_norm_dtype(dtype, x.dtype, "linalg.vector_norm")
+
+    computation_dtype, result_dtype = reduction_dtypes(
+        x, REDUCTION_OUTPUT_TYPE_KIND.COMPLEX_TO_FLOAT, dtype
+    )
+
+    to_result_dtype = partial(prims.convert_element_type, dtype=result_dtype)
+
+    # Implementation
+    if ord == 0.0:
+        return refs.sum(refs.ne(x, 0.0), dim=dim, keepdim=keepdim, dtype=result_dtype)
+    elif ord == float("inf"):
+        return to_result_dtype(refs.amax(prims.abs(x), dim=dim, keepdim=keepdim))
+    elif ord == float("-inf"):
+        return to_result_dtype(refs.amin(prims.abs(x), dim=dim, keepdim=keepdim))
+    else:
+        # From here on the computation dtype is important as the reduction is non-trivial
+        x = prims.convert_element_type(x, computation_dtype)
+        reduce_sum = partial(refs.sum, dim=dim, keepdim=keepdim)
+
+        # This choice should be made in ref
+        def fast_pow(x, p):
+            if p == 1.0:
+                return x
+            elif p == 2.0:
+                return x * x
+            elif p == 0.5:
+                return prims.sqrt(x)
+            else:
+                return prims.pow(x, p)
```

Comment:
Prefer calling torch operators when available, then refs if the ref has more functionality and the functionality is needed

prims should only be called when absolutely necessary (almost always by the operation that actually wraps the prim)

Reply:
Yes, sorry, this PR has been hanging here for way too long. I'll update it.